src/trs/rewrite.rs

```diff
@@ -0,0 +1,128 @@
+//! Hindley-Milner Typing for First-Order Term Rewriting Systems (no abstraction)
+//!
+//! Much thanks to:
+//! - https://github.com/rob-smallshire/hindley-milner-python
+//! - https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system
+//! - (TAPL; Pierce, 2002, ch. 22)
+
+use polytype::{Context, TypeSchema};
+use rand::Rng;
+use std::f64::NEG_INFINITY;
+use std::fmt;
+use term_rewriting as trs;
+
+use super::trace::Trace;
+use super::{Lexicon, SampleError, TypeError};
+
+/// Manages the semantics of a term rewriting system.
+#[derive(Debug, PartialEq, Clone)]
+pub struct RewriteSystem {
+    // TODO: may also want to track background knowledge here.
+    pub(crate) lex: Lexicon,
+    pub(crate) trs: trs::TRS,
+    pub(crate) ctx: Context,
+}
+impl RewriteSystem {
+    pub fn new(lex: &Lexicon, trs: trs::TRS) -> Result<RewriteSystem, TypeError> {
+        let lex = lex.clone();
+        let mut ctx = Context::default();
+        {
+            let lex = lex.0.read().expect("poisoned lexicon");
+            lex.infer_trs(&trs, &mut ctx)?;
+        }
```

Comment:
This scope could be collapsed into a one-liner, right, or does it get too long?

Reply:
I was wary of this because of how the lock guards get dropped at the end of scope, but if you don't assign it to a variable then it only lives as long as the statement. TIL!